ThirdpartyToolchain.cmake

```diff
@@ -0,0 +1,205 @@
+# Copyright (c) Facebook, Inc. and its affiliates.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# MODULE:   ThirdpartyToolchain
+#
+# PROVIDES:
+#   resolve_dependency( DEPENDENCY_NAME dependencyName
+#                    [REQUIRED_VERSION required version]
+#                    ...
+#   )
+#
+#       Provides the ability to resolve third party dependencies. If the dependency is
+#       already available in the system it will be used.
+#
+#       The DEPENDENCY_NAME argument is required. The dependencyName value will be used to search for
+#       the installed dependencies Config file and thus this name should match find_package()
+#       standards.
+#
+# EXAMPLE USAGE:
+#   # Download and setup or use already installed dependency.
+#   include(ThirdpartyToolchain)
+#   resolve_dependency(folly)
+#
+#
+#========================================================================================
+
+
+add_custom_target(toolchain)
+include(ExternalProject)
+
+#=====================================FOLLY==============================================
+
+if(DEFINED ENV{VELOX_FOLLY_URL})
+  set(FOLLY_SOURCE_URL "$ENV{VELOX_FOLLY_URL}")
+else()
+  set(VELOX_FOLLY_BUILD_VERSION v2022.07.11.00)
+  set(FOLLY_SOURCE_URL
+      "https://github.com/facebook/folly/archive/${VELOX_FOLLY_BUILD_VERSION}.tar.gz")
+  set(VELOX_FOLLY_BUILD_SHA256_CHECKSUM b6cc4082afd1530fdb8d759bc3878c1ea8588f6d5bc9eddf8e1e8abe63f41735)
+endif()
+
+macro(build_folly)
+  message(STATUS "Building Folly from source")
+  set(FOLLY_PREFIX "${CMAKE_CURRENT_BINARY_DIR}/folly-install")
+  set(FOLLY_INCLUDE_DIR "${FOLLY_PREFIX}/include")
+  file(MAKE_DIRECTORY ${FOLLY_INCLUDE_DIR})
+
+  set(FOLLY_STATIC_LIB
+      "${FOLLY_PREFIX}/lib/${CMAKE_STATIC_LIBRARY_PREFIX}folly${CMAKE_STATIC_LIBRARY_SUFFIX}"
+  )
+  set(FOLLY_BENCHMARK_STATIC_LIB
+      "${FOLLY_PREFIX}/lib/${CMAKE_STATIC_LIBRARY_PREFIX}follybenchmark${CMAKE_STATIC_LIBRARY_SUFFIX}"
+  )
+  # TODO Correctly use CMAKE_CXX_FLAGS defined on the project. Using them makes folly linking
+  # issues as described on:
+  # https://github.com/facebook/folly/blob/main/folly/container/detail/F14IntrinsicsAvailability.h#L32-L34
+  set(CMAKE_CXX_FLAGS_FOLLY "-mavx2 -mfma -mavx -mf16c -mlzcnt -std=c++17")
```

Comment:
These flags should be pulled from setup-helper-functions.sh; See here for an example on how to pull these in : https://github.com/facebookincubator/velox/blob/main/CMakeLists.txt#L168 .

Reply:
Thanks! Done!